### PR TITLE
Sample project to show the grapher capabilities

### PR DESCRIPTION
**Pull request by AndreaCrotti**
Add a sample project that uses the topology-grapher to render graphs from jackdaw and non jackdaw topologies.

### CODE REVIEW

**Comment by aviflax** [comment]
Is this ready for review? I don’t see any additions to the docs to point people to the new sample project.

**Comment by AndreaCrotti** [comment]
Not yet sorry, I'll add some more documentation and clean it up

**Comment by aviflax** [comment]
No problem, but maybe prefix the title with WIP when opening a PR for preliminary review or just for visibility?

File: README.md [hunk 1]
```diff
@@ -1,11 +1,76 @@
...
+```
+
+The `generate-zip` function is what takes care of transforming
+topologies objects into EDN files that are written out to disk.
+
+The important thing is that you need to be have a function that
+returns a topology object *without* having to actually start Kafka.
+
+This is normally fine, but if you are using something like the Stuart
+Sierra components library for example you will need to refactor the
```
Comment (aviflax) [inline README.md]:
```suggestion
Sierra [component](https://github.com/stuartsierra/component) for example you will need to refactor the
```

File: README.md [hunk 2]
```diff
@@ -1,11 +1,76 @@
...
-
 [![Clojars Project](https://img.shields.io/clojars/v/fundingcircle/topology-grapher.svg)](https://clojars.org/fundingcircle/topology-grapher)
 [![CircleCI](https://circleci.com/gh/FundingCircle/topology-grapher/tree/master.svg?style=svg)](https://circleci.com/gh/FundingCircle/topology-grapher/tree/master)
 [![codecov](https://codecov.io/gh/FundingCircle/topology-grapher/branch/master/graph/badge.svg)](https://codecov.io/gh/FundingCircle/topology-grapher)
 
+# Topology Grapher
+
 This library provides a means to build a directed graph (data) from a kafka streams topology.
 
+# How to integrate your application
```
Comment (aviflax) [inline README.md]:
```suggestion
## How to integrate into your application
```

I think it‘s best for Markdown docs to have only a single h1 header.

File: README.md [hunk 3]
```diff
@@ -1,11 +1,76 @@
...
+
+If you want to play around with this library you can check the [sample project](./sample_project/README.md).
+
+There are a few simple steps to follow to integrate your kafka streaming app.
+
+## 1. implement a command to call the describe-topologies
```
Comment (aviflax) [inline README.md]:
```suggestion
### 1. Create a namespace or function that invokes `generate-zip`
```

File: README.md [hunk 4]
```diff
@@ -1,11 +1,76 @@
...
+topologies objects into EDN files that are written out to disk.
+
+The important thing is that you need to be have a function that
+returns a topology object *without* having to actually start Kafka.
+
+This is normally fine, but if you are using something like the Stuart
+Sierra components library for example you will need to refactor the
+initialisation to avoid needing the full system started.
+
+## 2. add a call to that command in CircleCI
```
Comment (aviflax) [inline README.md]:
```suggestion
### 2. Create a CI job that invokes that namespace or function
```

File: README.md [hunk 5]
```diff
@@ -1,11 +1,76 @@
...
+
+This is an example of configuration you need to add to CircleCI to generate and publish graph information.
```
Comment (aviflax) [inline README.md]:
```suggestion
Here’s an example [CircleCI](https://circleci.com/) job that generates and publishes the graph data:
```

File: README.md [hunk 6]
```diff
@@ -1,11 +1,76 @@
...
+
+```
```
Comment (aviflax) [inline README.md]:
```suggestion
```yaml
```

File: README.md [hunk 7]
```diff
@@ -1,11 +1,76 @@
...
+      - restore_cache:
+          key: your-project-{{ checksum "project.clj" }}
+
+      # the precise command depends on how you integrated
+      - run: lein run describe-topologies
+      # not the real bucket yet since we are still waiting for one
+      # - run: aws s3 cp --recursive /tmp/graphs/$your-project-name s3://$your-bucket
+```
+
+This particular CircleCI step can run in parallel with any other job, and should not be required as dependency.
```
Comment (aviflax) [inline README.md]:
```suggestion
This CircleCI job can run in parallel with any other job.
```

I don’t think we need to provide quite so much detailed guidance on integrating with CircleCI.

File: README.md [hunk 8]
```diff
@@ -1,11 +1,76 @@
...
+
+And you are done, if you integrated in a way that all the topologies will always be published automatically
+there is 0 maintenance effort needed from now on.
```
Comment (aviflax) [inline README.md]: This paragraph seems unnecessary.

File: README.md [hunk 9]
```diff
@@ -1,11 +1,76 @@
...
+
+# Graph implementation
```
Comment (aviflax) [inline README.md]:
```suggestion
## Graph implementation
```

I think it‘s best for Markdown docs to have only a single h1 header.

File: README.md [hunk 10]
```diff
@@ -23,7 +88,7 @@ on the graph data. The intended flow of operation is:
 The graph generation has been tested on Kafka 1.0+ streams apps, both using the
 higher level DSL and lower level streams APIs.
 
-# Format of the Graph Data
+## Internal graph format
```
Comment (aviflax) [inline README.md]:
```suggestion
### Internal graph format
```

I think it‘s best for Markdown docs to have only a single h1 header.

File: README.md [hunk 11]
```diff
@@ -133,7 +198,7 @@ dependencies.
 * `lein test`
 * `lein install` (to install a local snapshot for any local integration testing)
 
-## Release
+# Release
```
Comment (aviflax) [inline README.md]:
```suggestion
## Release
```

I think it‘s best for Markdown docs to have only a single h1 header.

File: README.md [hunk 12]
```diff
@@ -142,67 +207,6 @@ To release a new version you just need to:
...
-      # - run: aws s3 cp --recursive /tmp/graphs/$your-project-name s3://$your-bucket
-```
-
-This particular CircleCI step can run in parallel with any other job, and should not be required as dependency.
-
-And you are done, if you integrated in a way that all the topologies will always be published automatically
-there is 0 maintenance effort needed from now on.
-
-## Contributing
+# Contributing
```
Comment (aviflax) [inline README.md]:
```suggestion
## Contributing
```

I think it‘s best for Markdown docs to have only a single h1 header.

File: README.md [hunk 13]
```diff
@@ -1,11 +1,76 @@
...
+initialisation to avoid needing the full system started.
+
+## 2. add a call to that command in CircleCI
+
+This is an example of configuration you need to add to CircleCI to generate and publish graph information.
+
+```
+  publish_topologies_graphs:
+    docker:
+      - image: *lein
```
Comment (aviflax) [inline README.md]: This `*lein` value seems odd to me. What does it mean? Is it a valid value or a placeholder of some kind?

File: deps.edn [hunk 14]
```diff
@@ -0,0 +1,6 @@
+{:deps
```
Comment (aviflax) [inline deps.edn]: I like tools-deps and `deps.edn` in general, but I’m just wondering why this is included in this PR? What does this have to do with the goal of this branch?

File: sample_project/deps.edn [hunk 15]
```diff
@@ -0,0 +1,11 @@
+{:deps
+ {fundingcircle/jackdaw {:mvn/version "0.6.9"}
+  org.clojure/clojure {:mvn/version "1.10.1"}
+  hawk {:mvn/version "0.2.11"}
+  aysylu/loom {:mvn/version "1.0.2"}
+  fundingcircle/topology-grapher {:local/root ".."}}
+
+ :mvn/repos
+ {"confluent" {:url "https://packages.confluent.io/maven/"}}
```
Comment (aviflax) [inline sample_project/deps.edn]: I guess this is needed because Jackdaw depends on packages from this repo? Whatever the reason, a comment explaining it would be helpful.

File: sample_project/deps.edn [hunk 16]
```diff
@@ -0,0 +1,11 @@
+{:deps
+ {fundingcircle/jackdaw {:mvn/version "0.6.9"}
+  org.clojure/clojure {:mvn/version "1.10.1"}
+  hawk {:mvn/version "0.2.11"}
+  aysylu/loom {:mvn/version "1.0.2"}
+  fundingcircle/topology-grapher {:local/root ".."}}
```
Comment (aviflax) [inline sample_project/deps.edn]: Since the sample project has code that directly references classes under `org.apache.kafka.streams` etc, I suggest adding Kafka Streams as an explicit top-level dependency here.

File: README.md (hunk 13 above)
Comment (AndreaCrotti) [reply README.md]: it's a variable, but yeah confusing, I'll use a real image example instead

File: sample_project/deps.edn (hunk 15 above)
Comment (AndreaCrotti) [reply sample_project/deps.edn]: It is actually related, since the sample project uses `deps.edn` to be able to build locally I needed a `deps.edn` here as well.

File: sample_project/deps.edn (hunk 16 above)
Comment (AndreaCrotti) [reply sample_project/deps.edn]: Yeah well they are transitive dependencies of jackdaw, but I guess it doesn't harm to add the explicit dependencies as well

File: sample_project/deps.edn (hunk 15 above)
Comment (aviflax) [reply sample_project/deps.edn]: Aha, makes sense!
